mdtk/degradations.py

```diff
@@ -514,16 +582,146 @@ def add_note(excerpt, min_pitch=MIN_PITCH, max_pitch=MAX_PITCH,
+    """
+    if excerpt.note_df.shape[0] == 0:
+        warnings.warn('WARNING: No notes to split. Returning None.',
+                      category=UserWarning)
+        return None
+    
+    # Find all splitable notes
+    valid_notes = []
+    
+    # Use indices because saving the df.loc objects creates copies
+    for note_index in range(excerpt.note_df.shape[0]):
+        if (excerpt.note_df.loc[note_index]['dur'] >=
+                (min_duration * (num_splits + 1))):
+            valid_notes.append(note_index)
+    
+    if not valid_notes:
+        warnings.warn('WARNING: No valid notes to split. Returning ' +
+                      'None.', category=UserWarning)
+        return None
+    
+    note_index = valid_notes[randint(len(valid_notes))]
+    
+    degraded = excerpt.copy()
+    
+    short_duration_float = (degraded.note_df.loc[note_index, 'dur'] /
+                            (num_splits + 1))
+    pitch = degraded.note_df.loc[note_index, 'pitch']
+    track = degraded.note_df.loc[note_index, 'track']
+    this_onset = degraded.note_df.loc[note_index, 'onset']
+    next_onset = this_onset + short_duration_float
+    
+    # Shorten original note
+    degraded.note_df.loc[note_index]['dur'] = int(round(short_duration_float))
+    
+    # Add next notes (taking care to round correctly)
+    for i in range(num_splits):
+        this_onset = next_onset
+        next_onset += short_duration_float
+        degraded.note_df = degraded.note_df.append(
+            {
+                'pitch': pitch,
+                'onset': int(round(this_onset)),
+                'dur': int(round(next_onset)) - int(round(this_onset)),
+                'track': track
+            },
+            ignore_index=True
+        )
+    
+    return degraded
 
 
 
 @set_random_seed
-def join_notes(excerpt, params):
-    """Combine two notes into one note."""
-    raise NotImplementedError()
+def join_notes(excerpt, max_gap=50):
+    """
+    Combine two notes of the same pitch and track into one.
+    
+    Parameters
+    ----------
+    excerpt : Composition
+        A Composition object of an excerpt from a piece of music.
+        
+    max_gap : int
+        The maximum gap length, in ms, for 2 notes to be able to be joined.
+        (They must always share the same pitch and track).
+        
+    seed : int
+        A seed to be supplied to np.random.seed(). Defaults to None, which
+        leaves numpy's random state unchanged.
+        
+    Returns
+    -------
+    degraded : Composition
+        A copy of the given excerpt, with one note split.
+    """
+    if excerpt.note_df.shape[0] < 2:
+        warnings.warn('WARNING: No notes to join. Returning None.',
+                      category=UserWarning)
+        return None
+    
+    valid_notes = []
+    
+    for _, track_df in excerpt.note_df.groupby('track'):
+        for _, pitch_df in track_df.groupby('pitch'):
+            notes = list(pitch_df.iterrows())
```

Comment:
This double groupby is going to be very slow - this is because it will split out into many groups (there are lots of pitches). It could be ok here because you're not using it within an apply (this was the issue with my csv non overlapping pitches validity check in the data_structures module.

I'm just noting this here, let's leave it for now.